pkg/controller/node/nodecontroller.go

```diff
@@ -482,37 +487,98 @@ func (nc *NodeController) monitorNodeStatus() error {
 			}
 		}
 	}
+	nc.handleDisruption(zoneToNodeConditions, nodes)
 
+	return nil
+}
+
+func (nc *NodeController) handleDisruption(zoneToNodeConditions map[string][]*api.NodeCondition, nodes *api.NodeList) {
+	newZoneStates := map[string]zoneState{}
 	for k, v := range zoneToNodeConditions {
 		newState := nc.computeZoneStateFunc(v)
-		if newState == nc.zoneStates[k] {
-			continue
+		newZoneStates[k] = newState
+		if _, had := nc.zoneStates[k]; !had {
+			nc.zoneStates[k] = stateInitial
 		}
-		if newState == stateFullSegmentation {
-			glog.V(2).Infof("NodeController is entering network segmentation mode in zone %v.", k)
-		} else if newState == stateNormal {
-			glog.V(2).Infof("NodeController exited network segmentation mode in zone %v.", k)
+	}
+
+	allWasFullyDisrupted := true
+	for _, v := range nc.zoneStates {
+		if v != stateFullDisruption {
+			allWasFullyDisrupted = false
+			break
 		}
-		for i := range nodes.Items {
-			if utilnode.GetZoneKey(&nodes.Items[i]) == k {
-				if newState == stateFullSegmentation {
-					// When zone is fully segmented we stop the eviction all together.
-					nc.cancelPodEviction(&nodes.Items[i])
-				}
-				if newState == stateNormal && nc.zoneStates[k] == stateFullSegmentation {
-					// When exiting segmentation mode update probe timestamps on all Nodes.
-					now := nc.now()
-					v := nc.nodeStatusMap[nodes.Items[i].Name]
-					v.probeTimestamp = now
-					v.readyTransitionTimestamp = now
-					nc.nodeStatusMap[nodes.Items[i].Name] = v
-				}
-			}
+	}
+	allAreFullyDisrupted := true
+	for _, v := range newZoneStates {
+		if v != stateFullDisruption {
+			allAreFullyDisrupted = false
+			break
 		}
-		nc.zoneStates[k] = newState
 	}
 
-	return nil
+	// Something either was responding in previous pass or is responding now. Semantics is as follows:
```

Comment:
Maybe something like this instead: "At least one node was responding in previous pass and in current pass" ?